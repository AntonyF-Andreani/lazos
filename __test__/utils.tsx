import React from "react";
import { render } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";

import theme from "theme";

/**
 * Colocar todos los providers, se puede mockear redux y las funciones
 */
const AllTheProviders: React.FC = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement) => render(ui, { wrapper: AllTheProviders });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
