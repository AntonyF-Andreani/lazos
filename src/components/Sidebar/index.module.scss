.listContainer {
  max-width: 100vw;
  width: 355px;

  & > ul {
    margin-top: 1em;
    margin-inline: 15px;

    & > li:hover {
      color: var(--primary);
    }

    & > li:first-of-type {
      margin-bottom: 20px;
    }
  }
}

.mediaContainer {
  display: flex;
  margin-top: auto;
  text-align: center;
  justify-content: center;
  gap: 10px;
  flex-direction: column;
  padding-block: 35px;

  & > p {
    text-transform: uppercase;
    color: #777;
    line-height: 26px;
    font-size: 0.8em;
    font-weight: 400;
  }
}

.media {
  display: flex;
  justify-content: center;
  gap: 0.3em;

  .mediaIcon {
    padding: 2px;
    width: 30px;
    height: 30px;
    vertical-align: middle;

    &:hover {
      border-radius: 50%;
      color: white;
    }
  }
}

.twitter:hover {
  background-color: #007bff;
}

.facebook:hover {
  background-color: #3b5a9a;
}

.linkedin:hover {
  background-color: #0073b2;
}

.youtube:hover {
  background-color: #c00;
}