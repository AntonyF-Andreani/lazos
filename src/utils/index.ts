import { ImageLoaderProps } from "next/image";

/**
 * Utilidad para extraerla URL con formatos optimizados para las imagenes del CMS (Servicio de Imagenes)
 * https://www.storyblok.com/docs/image-service
 */
export const StoryBlockImageURL = ({
  src,
  width,
  quality = 20,
  format = "webp",
  height = 0,
}: StoryBlockImageURLParams) => {
  return `${src}/m/${width}x${height}/filters:quality(${quality}):format(${format})`;
};

interface StoryBlockImageURLParams {
  /** URL de la imagen como la provee Storyblok */
  src: string;
  /** Width of image. The Storyblok API infers aspect ratio */
  width: number;
  /** Quality of image, for default its 20 (20%). */
  quality?: number;
  /** Format of image, for default its webp */
  format?: "webp" | "jpeg" | "png";
  /** Height of the image */
  height?: number;
}

export const StoryBlockLoader = ({ src, width, quality = 20 }: ImageLoaderProps) => {
  return StoryBlockImageURL({ src, width, quality });
};

/**
 * Util for join classes undefined or string and filter undefined
 * @param args classNames
 *
 *  @example const className = joinClasses("a", "b", undefined, "c") // "a b c"
 */
export const joinClasses = (...args: (string | undefined)[]): string => {
  return args.filter((x) => x).join(" "); // undefined is falsey
};

/**
 * Util to test diferrent viewports, must be used with jest.
 *
 * This util make a mock implementation of window.matchMedia
 *
 * @example prepareViewport("(max-width: 768px)")
 */
export const prepareViewport = (mediaQuery: string) => {
  Object.defineProperty(window, "watchMedia", {
    writable: true,
    configurable: true,
    value: mediaQuery,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    return {
      matches: query === mediaQuery,
      media: "",
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};
