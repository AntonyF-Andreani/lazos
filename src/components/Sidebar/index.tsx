import { faTimes } from "@fortawesome/pro-light-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Box, Drawer, Link, List, ListItem, ListItemIcon, Typography } from "@mui/material";
import React from "react";
import {
  faFacebookF,
  faLinkedinIn,
  faTwitter,
  faYoutube,
} from "@fortawesome/free-brands-svg-icons";

import SidebarItem, { SidebarItemAdapter } from "../SidebarItem";

import styles from "./index.module.scss";

import { joinClasses } from "utils";
import { Media, SidebarData } from "services/Cms/types";

export interface SidebarProps {
  open: boolean;
  onClose: () => void;
  data: SidebarData[];
  media?: Media;
}

const Sidebar = ({ open, onClose, data, media }: SidebarProps) => {
  return (
    <Drawer anchor="left" open={open} onClose={onClose}>
      <Box className={styles.listContainer} role="presentation">
        <List>
          <ListItem>
            <ListItemIcon onClick={onClose}>
              <FontAwesomeIcon cursor={"pointer"} icon={faTimes} size="2x" />
            </ListItemIcon>
          </ListItem>

          {data.length !== 0 &&
            data.map((item, i) => (
              <SidebarItem
                key={item._uid}
                data-testid={`sidebar-item-${item._uid}`}
                divider={i !== data.length - 1}
                {...SidebarItemAdapter(item)}
              />
            ))}
        </List>
      </Box>
      <Box className={styles.mediaContainer}>
        {media && (
          <Box className={styles.media}>
            {media.facebook && (
              <Link
                className={joinClasses(styles.mediaIcon, styles.facebook)}
                color="inherit"
                data-testid="sidebar-facebook"
                href={media.facebook}
              >
                <FontAwesomeIcon icon={faFacebookF} size="sm" />
              </Link>
            )}
            {media.twitter && (
              <Link
                className={joinClasses(styles.mediaIcon, styles.twitter)}
                color="inherit"
                data-testid="sidebar-twitter"
                href={media.twitter}
              >
                <FontAwesomeIcon icon={faTwitter} size="sm" />
              </Link>
            )}
            {media.linkedin && (
              <Link
                className={joinClasses(styles.mediaIcon, styles.linkedin)}
                color="inherit"
                data-testid="sidebar-linkedin"
                href={media.linkedin}
              >
                <FontAwesomeIcon icon={faLinkedinIn} size="sm" />
              </Link>
            )}
            {media.youtube && (
              <Link
                className={joinClasses(styles.mediaIcon, styles.youtube)}
                color="inherit"
                data-testid="sidebar-youtube"
                href={media.youtube}
              >
                <FontAwesomeIcon icon={faYoutube} size="sm" />
              </Link>
            )}
          </Box>
        )}
        <Typography>Â© {new Date().getFullYear()} Lazos</Typography>
      </Box>
    </Drawer>
  );
};

export default Sidebar;
