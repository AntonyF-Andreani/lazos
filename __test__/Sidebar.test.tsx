import React from "react";

import { render } from "./utils";

import Sidebar, { SidebarProps } from "components/Sidebar";

const defaultProps: SidebarProps = {
  open: true,
  onClose: () => {},
  data: [],
};

describe("<Sidebar />", () => {
  it("Should render", () => {
    const { asFragment } = render(<Sidebar {...defaultProps} />);

    expect(asFragment()).toMatchSnapshot();
  });
  it("Should render items", () => {
    const { getByTestId } = render(
      <Sidebar {...defaultProps} data={[{ _uid: "test", text: "text-test" }]} />
    );

    expect(getByTestId("sidebar-item-test")).toBeInTheDocument();
  });

  it("Should render media items", () => {
    const { getByTestId } = render(
      <Sidebar
        {...defaultProps}
        data={[{ _uid: "test", text: "text-test" }]}
        media={{
          linkedin: "linkedin",
          twitter: "twitter",
          facebook: "facebook",
          youtube: "youtube",
        }}
      />
    );

    expect(getByTestId("sidebar-linkedin")).toBeInTheDocument();
    expect(getByTestId("sidebar-twitter")).toBeInTheDocument();
    expect(getByTestId("sidebar-facebook")).toBeInTheDocument();
    expect(getByTestId("sidebar-youtube")).toBeInTheDocument();
  });
});
