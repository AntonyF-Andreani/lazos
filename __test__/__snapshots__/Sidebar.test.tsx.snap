// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> Should render 1`] = `<DocumentFragment />`;
